source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,704 @@
+
+  // Copy each basic block in the loop, give them new ids, and save state
+  // information.
+  for (const ir::BasicBlock* itr : basic_blocks) {
+    CopyBasicBlock(loop, itr, false);
+  }
+
+  // Set the previous continue block to point to the new header.
+  ir::Instruction& continue_branch = *state_.previous_continue_block_->tail();
+  continue_branch.SetInOperand(0, {state_.new_header_block->id()});
+
+  // As the algorithm copies the original loop blocks exactly, the tail of the
+  // latch block on iterations after the first one will be a branch to the new
+  // header and not the actual loop header. The last continue block in the loop
+  // should always be a backedge to the global header.
+  ir::Instruction& new_continue_branch = *state_.new_continue_block->tail();
+  new_continue_branch.SetInOperand(0, {loop->GetHeaderBlock()->id()});
+
+  // Update references to the old phi node with the actual variable.
+  const ir::Instruction* induction = loop_induction_variable_;
+  state_.new_inst[induction->result_id()] = GetPhiVariableID(
+      state_.previous_phi_, state_.previous_continue_block_->id());
+
+  if (eliminate_conditions &&
+      state_.previous_condition_block_ != loop_condition_block_) {
+    FoldConditionBlock(state_.previous_condition_block_, 1);
+  }
+
+  // Only reference to the header block is the backedge in the latch block,
+  // don't change this.
+  state_.new_inst[loop->GetHeaderBlock()->id()] = loop->GetHeaderBlock()->id();
+
+  for (auto& pair : state_.new_blocks) {
+    RemapOperands(pair.second);
+  }
+
+  // Swap the state so the new is now the previous.
+  state_.NextIterationState();
+}
+
+uint32_t LoopUnrollerUtilsImpl::GetPhiVariableID(const ir::Instruction* phi,
+                                                 uint32_t label) const {
+  for (uint32_t operand = 3; operand < phi->NumOperands(); operand += 2) {
+    if (phi->GetSingleWordOperand(operand) == label) {
+      return phi->GetSingleWordOperand(operand - 1);
+    }
+  }
+
+  return 0;
+}
+
+void LoopUnrollerUtilsImpl::FoldConditionBlock(ir::BasicBlock* condition_block,
+                                               uint32_t operand_label) {
+  ir::Instruction& old_branch = *condition_block->tail();
+
+  uint32_t new_target = old_branch.GetSingleWordOperand(operand_label);
+  std::unique_ptr<ir::Instruction> new_branch{
+      new ir::Instruction(ir_context_, SpvOp::SpvOpBranch, 0, 0,
+                          {{SPV_OPERAND_TYPE_ID, {new_target}}})};
+  ir_context_->KillInst(&old_branch);
+  condition_block->AddInstruction(std::move(new_branch));
+}
+
+void LoopUnrollerUtilsImpl::CloseUnrolledLoop(ir::Loop* loop) {
+  // Remove the OpLoopMerge instruction from the function.
+  ir::Instruction* merge_inst = loop->GetHeaderBlock()->GetLoopMergeInst();
+  ir_context_->KillInst(merge_inst);
+
+  // Remove the final backedge to the header and make it point instead to the
+  // merge block.
+  state_.previous_continue_block_->tail()->SetInOperand(
```

Comment:
That's specific to the loop you have been exposed to. You may need to transform the branch from conditional to unconditional.

Reply:
I now start to wonder if that's legal in shader spir-v...